### ISSUE TITLE
conversion pdf --> txt

### ISSUE THREAD

**Issue by ralessi** [maieul/ledmac#240]
Quand on cherche à convertir un fichier pdf produit avec eledmac vers un fichier .txt à l'aide de l'utilitaire "pdftotext" fourni avec la librairie (poppler)[http://poppler.freedesktop.org/], tout est bien interprété, y compris les passages en grec ancien. L'arabe non vocalisé passe aussi, mais l'arabe vocalisé ne passe pas, ce qui est très gênant.

Voici un mwe:
````\documentclass{article}

\usepackage{fontspec}
\usepackage{libertine}

\pagestyle{empty}

\usepackage{eledmac}
\footparagraph{A}
\nolemmaseparator[A]
\inplaceoflemmaseparator[A]{.5em}

\usepackage{arabxetex}

\begin{document}

L'arabe non vocalisé résiste bien à la conversion
pdf$\rightarrow$text. Mais l'arabe vocalisé ne passe
pas. Démonstration.

\beginnumbering

\pstart
\eledsection*{Arabe non vocalisé}
\pend

\pstart
Ἐρωτήματα· \edtext{εἰ}{\Afootnote{rest. Nikitas \textarab{hal}: εἴρεον
    V ἤρεον IRH}} ρήϊον ἀεὶ πληροῦσθαι ποτοῦ ἢ σιτοῦ.
\pend

\pstart
\eledsection*{Arabe vocalisé}
\pend

\pstart
Ἐρωτήματα· \edtext{εἰ}{\Afootnote{rest. Nikitas \textarab[voc]{hal}: εἴρεον
    V ἤρεον IRH}} ρήϊον ἀεὶ πληροῦσθαι ποτοῦ ἢ σιτοῦ.
\pend

\endnumbering

\end{document}
````
Voici le résultat de la commande
    pdftotext mwe.pdf

````L’arabe non vocalisé résiste bien à la conversion pdf→text. Mais l’arabe vocalisé
ne passe pas. Démonstration.

Arabe non vocalisé
Ἐρωτήματα· εἰ ρήϊον ἀεὶ πληροῦσθαι ποτοῦ ἢ σιτοῦ.

Arabe vocalisé
Ἐρωτήματα· εἰ ρήϊον ἀεὶ πληροῦσθαι ποτοῦ ἢ σιτοῦ.

2 εἰ rest. Nikitas ‫هل‬: εἴρεον V ἤρεον IRH 4 εἰ rest. Nikitas ‫ َهل‬: εἴρεον V ἤρεον IRH

````

**Comment by maieul** [comment]
a c'est un probl�me pdf2totext. Je croyais que tu parlais de l'outils que je suis en train de mettre en place directement depuis ledmac. 

Pour pdf2totext, je ne peux pas t'aider ;-).  Mais ils ont peut �tre leur propre syst�me de ticket.

**Comment by ralessi** [comment]
Pardon!
Je vais essayer avec l'outil que tu as fait, et je poste le résultat.

**Comment by maieul** [comment]
l'outil en question ne peux fonctionner qu'en lualatex. mais du coup, arabtex ne semble pas marcher en luatex. Apparement les gens tapent directement en unicode. voir
http://tex.stackexchange.com/q/21065/7712 et http://tex.stackexchange.com/q/171089/7712

**Comment by ralessi** [comment]
Oui, je connais... C'est bien pour cela que je n'utilise pas luatex ;)

**Comment by maieul** [comment]
alors mon outils ne pourra pas t'�tre utile, car il n�cessite imp�rativement luatex. J'explique le concept :
- ledmac divise les paragraphes en ligne.
- chaque ligne est stock� dans une boite.
- avec luatex, on r�cup�re chaque boite et on stocke dans fichier externes.

Donc peut �tre que des gens comp�tent en arabe ET en Tex pourraient faire arabluatex...

**Comment by ralessi** [comment]
arabluatex serait en effet une bonne idée. J'avais un peu regardé la chose, mais j'avais cru comprendre que teckit ne fonctionnait pas sous luatex. Mon souvenir est cependant assez vague, mais en tout cas, j'avais laissé tomber. Je devrais peut-être reprendre cette question, car je suis allé assez loin dans l'utilisation d'arabxetex que je connais presque à fond pour avoir corrigé pas mal d'erreurs de transcription et de rendu en arabe.

Quand j'utilise ton outil sous lualatex. j'obtiens cette erreur à chaque mot arabe:
````
! LuaTeX error [\directlua]:1: invalid escape sequence near '\l'.
\save@notecontent ... "..s.." ]#3\noexpand \n ") }
                                                   } 
````
Pourtant, le pdf est bon, avec de l'arabe vocalisé et non vocalisé. Mais eledpublish-concept.A est vide...

**Comment by maieul** [comment]
tu frappe comment l'arabe? en caract�re arabe ou autrement?
cela �tant c'est encore un "concept" ce projet. Pour le moment le grec est pas trop mal, l'hebreu passe bien dans le fichier text mais du coup on le perd dans le pdf (car les polices changent la valeur des caract�res unicodes)

**Comment by ralessi** [comment]
Sous lualatex, je l'ai bien sûr saisi en arabe unicode, comme ceci: هل (non vocalisé) et هَل (vocalisé). Si je force le passage sur les deux erreurs, le pdf est bon.

**Comment by maieul** [comment]
avec ce fichier https://github.com/maieul/ledmac/blob/d0ba45423b550cc291320bb91bd2e65b993149d4/eledpublish-concept.tex j'ai pas de souci, en tt cas pour les corps de texte (apr�s je ne sais pas effectivement comment ins�rer facilement du code tex dans du RTL !)

**Comment by ralessi** [comment]
Je vois!
Mais pour avoir mon arabe normalement dans le pdf sous lualatex, je fais ceci dans le préambule:
````
\newfontfamily\arabicfont[Script=Arabic, Scale=1.2]{Scheherazade}
\newcommand{\textarabic}[1]{\bgroup\luatextextdir TRT\arabicfont #1\egroup}
````
Ensuite, je mets les mots arabes comme ceci \textarabic{هل}.

Je viens de faire comme tu as fait. Cette fois, la compilation va jusqu'au bout, et le fichier eledpublish-concept.A est correct. Mais dans le pdf, les mots arabes ont disparu.

**Comment by maieul** [comment]
oui, c'est le m�me souci qu'avec l'hebreu alors. les polices doivent changer le caract�re unicode. du coup c'est soit l'un soit l'autre (voir http://tex.stackexchange.com/a/228324/7712 derni�res remarques)

>

**Comment by ralessi** [comment]
Maintenant je comprends. Je vais essayer de voir comment on pourrait faire un arabluatex...

**Comment by maieul** [comment]
�crit � Khaled Hosny, il pourrait te dire.


En ce qui concerne l'arabe tap� en unicode, voit le dernier commit sur la branche. Je pense que c'est ce que tu veux au final non ?

**Comment by maieul** [comment]
sinon, sur le principe tu pense que c'est la bonne solution pour s'arranger avec les �diteurs?

**Comment by ralessi** [comment]
Je vais écrire à Khaled Hosny.

Je viens de tester le dernier commit sur la branche, et tout est bon, à la fois dans eledpublish-concept.A et dans le pdf!

En soi, je pense que l'idée est excellente, car elle permet de fournir un document de travail précis. Cela suppose que l'éditeur travaille avec le pdf produit par eledmac sur la table, puis reprenne les lignes sorties par lualatex pour les adapter, en refaisant la numérotation, car les sauts de ligne auront changé. Cependant, quand on voit que certains éditeurs n'hésitent pas à demander aux auteurs de tout refaire avec CTE on peut se demander s'ils accepteront de faire le travail!

**Comment by maieul** [comment]
je ne sais pas. Il me semble que pour les gens qui font des �ditions en Word (si, si j'en connais !) l'�diteur accepte ce syst�me. 

Il faut que je me renseigne aupr�s d'un de mes coll�gues qui travaille pour le CCSA, mais il me semble que lorsqu'ils avaient voulu envoyer � l'�diteur le pdf de eledmac, celui-ci avait demand� un word.
je crois que je vais essayer de sortir une beta ce week-end. Maintenant que les principaux probl�mes sont pos�s. On verra les r�actions.

ps : tu verra qu'il y aussi un fichier .maintext qui est produit !

**Comment by ralessi** [comment]
Oui! Un Word, c'est ce qu'on me demande à moi! Ils refont tout ensuite. C'est la raison pour laquelle je testais pdftotext au départ. Donc, eledpublish sera le bienvenu pour ces éditeurs, je pense.

Je viens de voir le .maintext qui est parfait!

**Comment by ralessi** [comment]
Question: penses-tu qu'il faille se mettre à lualatex? Il semble que ce soit l'avenir, et que cet avenir se rapproche...

**Comment by maieul** [comment]
jusqu'� maintenant je ne voyais pas l'int�r�t par rapport � XeTeX. Maintenant je vois (et il n'y a pas de solution !).

Enfin on pourrait imaginer un eledpublisher simplif�, qui ne produirait pas le .maintext, mais ce serait pas top.

**Comment by ralessi** [comment]
Eh oui... c'est ce que je pense aussi. Mieux vaut anticiper!

Au fait: il existe un package qui s'appelle "dad" et qui permet de transcrire l'arabe pour luatex! Je viens de vérifier: ça marche, mais pas avec eledpublish! Peux tu vérifier de ton côté?

dad n'est pas aussi complet qu'arabxetex, mais je vais le scruter: je suis certain qu'on peut l'adapter aux besoins des philologues!

**Comment by maieul** [comment]
comme tjr, un MWE (surtout que je connais pas l'arabe !)

**Comment by ralessi** [comment]
Le même: seule différence: chargement de dad, et annulation des commandes que j'avais mises pour l'arabe sous luatex:
```
\documentclass{article}
\usepackage[none]{hyphenat}
\usepackage{fontspec}
\usepackage{libertine}

\usepackage{eledmac,eledpar}

\usepackage{dad}

\makeatletter
\def\print@line{
    \save@linecontent
    \affixpstart@num%
    \hb@xt@ \linewidth{\do@insidelinehook\l@dld@ta%
        \add@inserts\affixside@note%
        \l@dlsn@te
        {\ledllfill\hb@xt@ \wd\one@line{\new@line\inserthangingsymbol \l@dunhbox@line{\one@line}}\ledrlfill\l@drd@ta%
        \l@drsn@te
      }}%
    \add@penalties%
}

\def\save@linecontent{
	\setbox0=\copy\one@line
	%\eledmac@warning{here}
	\directlua{
	local char = unicode.utf8.char
	  local function nodeText(n)
	    local t =  {}
	    for x in node.traverse(n) do
	      % glyph node
	      if x.id == 37 then
	        table.insert(t,char(x.char))
	      % glue node
	      elseif x.id == 10 and x.subtype == 0 then
	        table.insert(t," ")
	      end
	    end
	    return table.concat(t)
	  end
	  local n = tex.getbox(0)
	  f:write("\the\line@num\space : "..nodeText(n.head),"\noexpand\n")
	}
}
\Xlemmadisablefontselection[A]
\newcommand{\save@notecontent}[4]{
	   \def\extractline@##1|##2|##3|##4|##5|##6|##7|{##2}%
   \def\extractsubline@##1|##2|##3|##4|##5|##6|##7|{##3}%
   \def\extractendline@##1|##2|##3|##4|##5|##6|##7|{##5}%
   \def\extractendsubline@##1|##2|##3|##4|##5|##6|##7|{##6}%
  {\let\edtext\dummy@edtext%
  \let\no@expands\relax
  \edef\tag{#2}
  \edef\content{\unexpanded{#3}}
  \patchcmd{\tag}{\no@expands}{}{}{}{}
\def\ignore##1{##1}
\let\textarabic\ignore
	\directlua{ 
 l="\luaescapestring{\tag}"
 t="\luaescapestring{\content}"
 n:write("\extractline@#1|-\extractendline@#1| "..l.." ]"..t.."\noexpand\n")
	}
	}
}

\let\old@normalfootfmt\normalfootfmt
\renewcommandx{\normalfootfmt}[4][4=Z]{% 4th arg is optional, for backward compatibility
  \save@notecontent{#1}{#2}{#3}{#4}
  \ledsetnormalparstuff%
  \hangindent=\csuse{Xhangindent@#4}
  \strut{\printlinefootnote{#1}{#4}}%
  {\nottoggle{Xlemmadisablefontselection@#4}{\select@lemmafont#1|#2}{#2}}%
  \iftoggle{nosep@}{\hskip\csuse{inplaceoflemmaseparator@#4}}{\ifcsempty{lemmaseparator@#4}%
    {\hskip\csuse{inplaceoflemmaseparator@#4}}%
    {\nobreak\hskip\csuse{beforelemmaseparator@#4}\csuse{lemmaseparator@#4}\hskip\csuse{afterlemmaseparator@#4}\relax%
  }}%
  #3\strut\par}


	 \directlua{
	   f = io.open("\jobname.maintext","w") f:write(" ")
	   n=io.open("\jobname.A","w")
	    n:write(" ")
	    }
\begin{document}

L'arabe non vocalisé résiste bien à la conversion
pdf$\rightarrow$text. Mais l'arabe vocalisé ne passe
pas. Démonstration.

\beginnumbering

\pstart
\eledsection*{Arabe non vocalisé}
\pend

\pstart
Ἐρωτήματα· \edtext{εἰ}{\Afootnote{rest. Nikitas \arab{هل}: εἴρεον
    V ἤρεον IRH}} ρήϊον ἀεὶ πληροῦσθαι ποτοῦ ἢ σιτοῦ.
\pend

\pstart
\eledsection*{Arabe vocalisé}
\pend

\pstart
Ἐρωτήματα· \edtext{εἰ}{\Afootnote{rest. Nikitas \arab{هَل}: εἴρεον
    V ἤρεον IRH}} ρήϊον ἀεὶ πληροῦσθαι ποτοῦ ἢ σιτοῦ.
\pend

\endnumbering

 \directlua{f:close() 
 n:close()
 }


\end{document}
````

**Comment by ralessi** [comment]
En tout cas, je suis très enthousiaste. Je vais passer la journée de demain sur dad pour bien comprendre comment il fonctionne et pour l'adapter.

**Comment by maieul** [comment]
c'est un probl�me en fait assez simple : les commandes de bascule de polices (en l'occurence \textarabic ou \arab) sont "expand�" (c'est du voc TeX) lors de la sortie.

La solution c'est juste de changer leur comportement dans la fonction de sorti.

Pour le moment cela passe par la ligne \let\arab\ignore, mais � terme cela sera une configuration de eledpublisher, quand le code sera d�grossi.

A propose de config, je me demandais ce qu'il valait mieux faire : une config en LaTeX ou une config en Lua ?

**Comment by maieul** [comment]
a oui, sinon je vais appeler cela eledpub, ce sera plus cohérent avec les nominations.

**Comment by maieul** [comment]
(pour ton pb, voir la nouvelle version)

**Comment by ralessi** [comment]
Bien sûr, suis-je bête. Pour la config, en lua je pense. C'est plus logique. Je vais aller voir la dernière version.

**Comment by maieul** [comment]
le truc pour la config, c'est que cela veut dire apprendre aux gens un syntaxe (d'ailleurs moi aussi je dois l'apprendre !)

**Comment by ralessi** [comment]
Je n'y avais pas pensé. Donc peut-être en LaTeX pour le moment...

**Comment by maieul** [comment]
bon, apr�s la syntaxe ce ne serait pas super compliqu� : 

variable = "valeur"
tableaux = {"variable", "variable"}

**Comment by ralessi** [comment]
En effet: les utilisateurs pourront s'adapter.

Avec la dernière version, j'ai ceci:
````
ERROR: Font \OT1/dad/m/n/8=dad at 9.59998pt not loadable: metric data not found or bad
````
Pourtant, avec dad+lualatex+eledmac, un mwe passe sans problème.

**Comment by maieul** [comment]
oui, j'ai cela aussi. Je ne sais pas. Je pense que c'est les notes mais � voir.

**Comment by maieul** [comment]
le pb n'esz pas li� a eledpublish. Je viens de commenter enti�rement le code

**Comment by ralessi** [comment]
étrange, car un mwe avec \arab dans les notes passe très bien. je n'ai pas compris. Je suis en train de tenter de me familiariser avec dad.

**Comment by ralessi** [comment]
Le probl�me vient d'une incompatibilit� entre dad et fontspec. Je
continue � chercher.
